server/server.go

```diff
@@ -0,0 +1,472 @@
+	tlsConn := tls.Server(conn, s.TLSConfig.Clone())
+
+	c := &eppConn{conn: tlsConn, maxMessageSize: s.MaxMessageSize}
+
+	s.mu.Lock()
+	s.activeConn[c] = struct{}{}
+	s.mu.Unlock()
+
+	// Set up a cancel context that is passed to handlers so that they, if needed,
+	// can be notified when the connection shuts down.
+	ctx, cancelCtx := context.WithCancel(context.Background())
+
+	// Setup some cleanup for when the session exits.
+	defer func() {
+		_ = c.Close()
+
+		// No need to remember the closeChan anymore.
+		s.mu.Lock()
+		delete(s.activeConn, c)
+		s.mu.Unlock()
+
+		if s.CloseConnHook != nil {
+			s.CloseConnHook(ctx, c.conn.(*tls.Conn))
+		}
+
+		cancelCtx()
+
+		// Countdown the wait group so that the entire listener can shut down
+		// when this reaches zero if it wants to.
+		s.wg.Done()
+	}()
+
+	err := setDeadlines(c.conn, s.ReadTimeout, s.WriteTimeout)
+	if err != nil {
+		s.logError("handshake deadlines", err)
+		return
+	}
+
+	err = tlsConn.Handshake()
+	if err != nil {
+		s.logDebug("handshake", err)
+		return
+	}
+
+	if s.ConnContext != nil {
+		// This is where the user can set up any context data for the
+		// connection, for example userID's etc.
+		ctx, err = s.ConnContext(ctx, tlsConn)
+		if err != nil {
+			// We don't log the error here because it's only here to signal
+			// that the user wanted to abort the connection.
+			return
+		}
+	}
+
+	// The responseWriter can be reused for each command.
+	rw := epplib.ResponseWriter{}
+
+	err = setDeadlines(c.conn, s.ReadTimeout, s.WriteTimeout)
+	if err != nil {
+		s.logError("set deadlines for greeting", err)
+		return
+	}
+
+	// We have properly connected so we need to begin by sending the greeting.
+	s.Greeting(ctx, &rw)
+
+	err = rw.FlushTo(c.conn)
+	if err != nil {
+		s.logError("flush greeting", err)
+		return
+	}
+
+	if rw.ShouldCloseAfterWrite() {
+		return
+	}
+
+	maxDeadline := deadlineFromTimeout(s.Timeout)
+
+	for {
+		deadline := getClosestDeadline(
+			maxDeadline,
+			deadlineFromTimeout(s.IdleTimeout),
+		)
+
+		err := c.conn.SetDeadline(deadline)
+		if err != nil {
+			s.logError("set deadlines for await message", err)
+			return
+		}
+
+		// Wait for a message to appear on the connection.
+		cmd, err := c.AwaitMessage()
+		if err != nil { // nolint:nestif // Needed.
+			if errors.Is(err, os.ErrDeadlineExceeded) {
+				// We have reached the deadline for this session, we now need
+				// to disconnect.
+				return
+			}
+
+			if errors.Is(err, net.ErrClosed) ||
+				errors.Is(err, syscall.ECONNRESET) ||
+				errors.Is(err, syscall.EPIPE) {
+				// Either we have closed AwaitMessage or the client has closed
+				// the connection.
+				return
+			}
+
+			if errors.Is(err, io.EOF) {
+				// The client has closed the connection.
+				return
+			}
+
+			if errors.Is(err, io.ErrUnexpectedEOF) {
+				// We don't want to turn this of entirely
+				s.logInfo("await message", err)
+				return
+			}
+
+			if strings.Contains(err.Error(), "message size exceeds limit") {
```

Comment:
Should this be a defined error instead of using `strings.contains`?

Reply:
We have one defined, I'll update this check to errors.Is() instead.